### PR TITLE
OssFuzz: Do not read/use undefined values in UndoXYB

### PR DESCRIPTION
**Pull request by eustas**
Refactor code to extract common part. This allows to deal with MSAN in one place (instead of 6).

### CODE REVIEW

File: lib/jxl/common.h [hunk 1]
```diff
@@ -30,6 +30,15 @@
 namespace jxl {
 // Some enums and typedefs used by more than one header file.
 
+// Define MEMORY_SANITIZER if memory sanitizer is enabled.
```
Comment (deymo) [inline lib/jxl/common.h:33]: MEMORY_SANITIZER enables a bunch of other behavior which we should re-evaluate, in particular the initialization of padding, which may not be an issue with recent versions. This should probably be in a different commit.

Comment (eustas) [reply lib/jxl/common.h:33]: I could make a separate commit, but the fix won't work without this change.
IIUC MEMORY_SANITIZER macro is a legacy way, currently documentation says "In some cases one may need to execute different code depending on whether MemorySanitizer is enabled. __has_feature can be used for this purpose.". In brotli we do check both macro and has_feature to detect MSAN build.

Comment (eustas) [reply lib/jxl/common.h:33]: Or we could move this to compiler options.

File: lib/jxl/dec_reconstruct.cc [hunk 2]
```diff
@@ -68,9 +68,17 @@ Status UndoXYBInPlace(Image3F* idct, const Rect& rect,
...
+        const auto mask = Iota(d, x) < Set(d, rect.xsize());
+        const auto sentinel = Set(d, kSanitizerSentinel);
+        const auto in_opsin_x = IfThenElse(mask, Load(d, row0 + x), sentinel);
+        const auto in_opsin_y = IfThenElse(mask, Load(d, row1 + x), sentinel);
+        const auto in_opsin_b = IfThenElse(mask, Load(d, row2 + x), sentinel);
+#else
         const auto in_opsin_x = Load(d, row0 + x);
         const auto in_opsin_y = Load(d, row1 + x);
         const auto in_opsin_b = Load(d, row2 + x);
+#endif
```
Comment (deymo) [inline lib/jxl/dec_reconstruct.cc:81]: Can we poison the portion of row{0,1,2} + x in line 91?

Comment (eustas) [reply lib/jxl/dec_reconstruct.cc:81]: Likely, yes. Do you propose to unpoison-process-poison here?

Comment (deymo) [reply lib/jxl/dec_reconstruct.cc:81]: yes. The unpoison-process-poison here will be founded on the fact that XybToRgb uses the undefined values but it is safe to do so; but we shouldn't assume that the output xybtorgb is initialized in other places because of this.
A comment about why next to the #ifdef MEMORY_SANITIZER explaining why would be appreciated.

**Comment by eustas** [comment]
@deymo PTAL. In the next PR I'll add lib/sanitizers.h and after that apply unpoison-process-poison scheme to this code.

**Comment by eustas** [comment]
@deymo PTALl nota-bene, now MSAN fails at later stage, but I think PR is big enough, to split more fixes in other PR

File: lib/jxl/dec_reconstruct.cc [hunk 3]
```diff
@@ -37,141 +37,114 @@ HWY_BEFORE_NAMESPACE();
 namespace jxl {
 namespace HWY_NAMESPACE {
 
-Status UndoXYBInPlace(Image3F* idct, const Rect& rect,
+template <typename Op>
```
Comment (deymo) [inline lib/jxl/dec_reconstruct.cc:40]: I think there's some potential here to reduce the binary size by removing this template and using a virtual function. It won't be inlined of course but given that we only do a Store afterwards I don't think it will make much difference in runtime performance. I'll do this in a follow up.

**Review by deymo — APPROVED** [review]
